packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts

```diff
@@ -0,0 +1,55 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {DataSource} from 'loopback-datasource-juggler';
+import {Note} from '../../fixtures/models/note.model';
+import {NoteRepository} from '../../fixtures/repositories/note.repository';
+
+describe('add method to crud repository via mixin', () => {
+  let ds: DataSource;
+  let repo: NoteRepository;
+
+  before(async () => {
+    ds = new DataSource({
+      connector: 'memory',
+    });
+    repo = new NoteRepository(ds);
+  });
+
+  it('non-mixin methods exist', () => {
+    const crudMethods = [
+      'create',
+      'createAll',
+      'find',
+      'updateAll',
+      'deleteAll',
+      'count',
+      'save',
+      'update',
+      'delete',
+      'findById',
+      'updateById',
+      'replaceById',
+      'deleteById',
+      'exists',
+    ];
+    const methodsFound = crudMethods.filter(methodName => {
+      return methodName in repo;
+    });
```

Comment:
```ts
const methodsFound = crudMethods.filter(methodName => typeof repo[methodName] === 'function');
```

Reply:
> ```ts
> const methodsFound = crudMethods.filter(methodName => typeof repo[methodName] === 'function');
> ```

Cannot seem to use this...it gives an error:

```
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'NoteRepository'.
  No index signature with a parameter of type 'string' was found on type 'NoteRepository'.ts(7053)
```

Reply:
Re-posting from our chat:

Personally, I am not a big fan of checking exactly what methods an interface or a class has. Such tests are brittle, because when we add a new method to `DefaultCrudRepository`, then the test will start to fail. Surely the test is not interested in exact method names provided by `DefaultCrudRepository`? IIUC, you want to say "repo has all DefaultCrudRepository members plus `findByTitle`". So the correct solution is to list all members of DefaultCrudRepository, add `findByTitle` to that array, and use that as the expected list of members in `repo`.

---

Let's take a look at your code:

```ts
    const methodsFound = crudMethods.filter(methodName => {
      return methodName in repo && typeof repo[methodName] === 'function';
    });
```

It takes an array of strings (crudMethods) and for each string value, it tries to access a member of repo with that name.

The problem is that from TypeScript's point of view, those strings can be arbitrary - you can sneak in `explode` or make a typo and use `craete` instead of `create`.

When analyzing the expression `repo[methodName]`, TypeScript is not able to figure out what type does it resolves to. It's a valid statement in JavaScript and because TypeScript cannot figure out what's the type, it uses `any`.

The problem with `any` is that once you get `any`, all further compile checks are pretty much disabled. This is often difficult to spot, therefore we have configured the compiler to tell us about these situations and reject them with errors.

I am proposing to build the list of expected methods as follow (please add the typeof function check yourself):

```ts
    const crudMethods: (keyof DefaultCrudRepository<
      Note,
      typeof Note.prototype.id
    >)[] = Object.keys(DefaultCrudRepository);
```

With that change in place, the compiler will discover the problem described in my comment https://github.com/strongloop/loopback-next/pull/4723#pullrequestreview-367270278 below.

Reply:
The code snippet above does not work, here is the corrected version:

```ts
    const crudProto = DefaultCrudRepository.prototype;
    const crudMembers = Object.keys(crudProto) as (keyof DefaultCrudRepository<
      Note,
      typeof Note.prototype.id
    >)[];
    const crudMethods = crudMembers.filter(
      key => typeof crudProto[key] === 'function',
    );
```

Reply:
Good catch!

Reply:
@bajtos thanks for the workaround. But I decided only to test `create`, `findById`, and `findByTitle` ; instead of all the crud functions.

Reply:
@bajtos , actually after having a discussion with @raymondfeng , we decided to keep the using of `Model` instead of `Entity` for all the mixin functions, and use `CrudRepository` instead of `EntityCrudRepository` for the sole reason that the mixin method in the documentation example add a `findByTitle` method which only needs to use `repo.find`...it doesn't need `create` or `findById` nor `exists`. So our mocha tests will reflect this as well. I have reverted the repomixin changes, and will only test `create` and `findByTitle`.  :)

Reply:
```
    const crudProto = DefaultCrudRepository.prototype;
    const crudMembers = Object.keys(crudProto) as (keyof DefaultCrudRepository<
      Note,
      typeof Note.prototype.id
    >)[];
    const crudMethods = crudMembers.filter(
      key => typeof crudProto[key] === 'function',
    );
```
This just produces empty arrays for me.

Reply:
> we decided to keep the using of `Model` instead of `Entity` for all the mixin functions, and use `CrudRepository` instead of `EntityCrudRepository` for the sole reason that the mixin method in the documentation example add a `findByTitle` method which only needs to use `repo.find`...it doesn't need `create` or `findById` nor `exists`.

Sure, that makes a lot of sense. I my view, we are designing the mixing to require only those methods that we actually need in the mixin implementation. If `findByTitle` needs only the `find` method, then we should require `CrudRepository` only and don't ask for `EntityCrudRepository` methods that we don't need.

> will only test `create` and `findByTitle`. 

I am afraid I strongly disagree with the second part. When the user calls `FindByTitleRepositoryMixin` and provides a subclass of `DefaultCrudRepository` (which implements `EntityCrudRepository` interface), then the class created by the mixin should have all `EntityCrudRepository` methods. This will be always true at runtime, we need to find a way how to make the TypeScript compiler understand that at compile-time too.

A test calling `repo.findById` verifies both compile-time and run-time behavior, I think it's crucial to keep such test included.

Based on the newer comments and the the latest version presented in cbcc296, I think my concerns are actually addressed because the test for `findById` is present :+1: